hax/hax/motr/delivery.py

```diff
@@ -69,14 +69,45 @@ def wait_for_any(self,
                       self.recently_delivered[promise])
             del self.recently_delivered[promise]
 
+    def wait_for_all(self,
+                     promise: HaLinkMessagePromise,
+                     timeout_sec: float = 30.0):
+        """
+        Blocks the current thread until all of the messages in
+        promise._ids are reported by Motr as delivered.
+
+        Raises NotDelivered exception when timeout_sec exceeds.
+        """
+        for msg in promise._ids:
+            condition = Condition()
+            with self.lock:
+                self.waiting_clients[promise] = condition
+
+            with condition:
+                LOG.debug('Blocking until %s is confirmed', promise)
+                condition.wait(timeout=timeout_sec)
+            with self.lock:
+                if promise not in self.recently_delivered:
+                    raise NotDelivered('None of message tags =' +
+                                       str(promise) +
+                                       '  were delivered to Motr')
+                msgid = self.recently_delivered[promise]
+                if msgid in promise:
+                    promise._ids.remove(msgid)
```

Comment:
Not sure that this is needed. If we don't remove the id from the promise, the algorithm will keep working, but no CPU time will be spent on removal.

Reply:
Removed for now.